modules/host/network.go

```diff
@@ -49,6 +49,78 @@ func (h *Host) threadedUpdateHostname(closeChan chan struct{}) {
 	}
 }
 
+// threadedTrackWorkingStatus periodically checks if the host is working,
+// where working is defined as having received 3 settings calls in the past 15
+// minutes.
+func (h *Host) threadedTrackWorkingStatus(closeChan chan struct{}) {
+	defer close(closeChan)
+
+	for {
+		prevSettingsCalls := atomic.LoadUint64(&h.atomicSettingsCalls)
+		select {
+		case <-h.tg.StopChan():
+			return
+		case <-time.After(workingStatusFrequency):
+		}
+		settingsCalls := atomic.LoadUint64(&h.atomicSettingsCalls)
+
+		// sanity check
+		if prevSettingsCalls > settingsCalls {
+			build.Severe("the host's settings calls decremented")
+			continue
+		}
+
+		h.mu.Lock()
+		if settingsCalls-prevSettingsCalls >= workingStatusThreshold {
+			h.workingStatus = WorkingStatusWorking
+		} else {
+			h.workingStatus = WorkingStatusNotWorking
+		}
+		h.mu.Unlock()
+	}
+}
+
+// threadedTrackConnectabilityStatus periodically checks if the host is
+// connectable at its netaddress.
+func (h *Host) threadedTrackConnectabilityStatus(closeChan chan struct{}) {
+	defer close(closeChan)
+
+	for {
+		select {
+		case <-h.tg.StopChan():
+			return
+		case <-time.After(connectabilityCheckFrequency):
+		}
+
+		h.mu.RLock()
+		autoAddr := h.autoAddress
+		userAddr := h.settings.NetAddress
+		h.mu.RUnlock()
+
+		activeAddr := autoAddr
+		if userAddr != "" {
+			activeAddr = userAddr
+		}
+
+		dialer := &net.Dialer{
+			Cancel:  h.tg.StopChan(),
+			Timeout: connectabilityCheckTimeout,
+		}
+		conn, err := dialer.Dial("tcp", string(activeAddr))
+		if err != nil {
+			h.mu.Lock()
+			h.connectabilityStatus = ConnectabilityStatusNotConnectable
+			h.mu.Unlock()
+			continue
```

Comment:
continue here is after the `conn.Close()`, need to make sure you call close before you continue